lib/poseidon/sync_producer.rb

```diff
@@ -71,7 +72,8 @@ def send_messages(messages)
       end
 
       if messages_to_send.pending_messages?
-        raise "Poseidon: Failed to send all messages: #{messages_to_send.messages} remaining"
+        failed_messages = messages_to_send.messages
+        raise Errors::MessageFailedToSendError.new(failed_messages, "Poseidon: Failed to send #{failed_messages.size} out of #{total_message_count} messages")
```

Comment:
I *believe* partial exceptions can still occur that would result in you not triggering this exception.

For example, I believe `refresh_metadata` can throw exceptions -- so I guess the question would be, what would happen if a partial failure occurred and we then failed to refresh the metadata?

We might want to put a begin / rescue around a larger portion of this code.